AC Tool should be cluster-aware

(AEM 6.1, AC Tool 1.9.5)

When the AC tool is deployed on a cluster, the situation can occur that the AC Tool is started on more than 1 cluster node in parallel, thus leading to merge exceptions when persisting data to the repository (both ACLs and the history).

The AC Tool should run only a single cluster node; or when running on multiple cluster nodes, it should not run in parallel, thus avoiding exceptions like this; for me it is not clear what the impact of this exception is (we will probably run the tool again just to make sure).

```
23.08.2017 08:34:53.424 *ERROR* [OsgiInstallerImpl-ACTool-Config-Worker] biz.netcentric.cq.tools.actool.installationhistory.impl.AcHistoryServiceImpl RepositoryException:
javax.jcr.InvalidItemStateException: OakState0001: Unresolved conflicts in /var/statistics/achistory
                at org.apache.jackrabbit.oak.api.CommitFailedException.asRepositoryException(CommitFailedException.java:237)
                at org.apache.jackrabbit.oak.api.CommitFailedException.asRepositoryException(CommitFailedException.java:212)
                at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.newRepositoryException(SessionDelegate.java:661)
                at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.save(SessionDelegate.java:487)
                at org.apache.jackrabbit.oak.jcr.session.SessionImpl$8.performVoid(SessionImpl.java:424)
                at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.performVoid(SessionDelegate.java:266)
                at org.apache.jackrabbit.oak.jcr.session.SessionImpl.save(SessionImpl.java:421)
                at sun.reflect.GeneratedMethodAccessor52.invoke(Unknown Source)
                at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.lang.reflect.Method.invoke(Method.java:497)
                at org.apache.sling.jcr.base.SessionProxyHandler$SessionProxyInvocationHandler.invoke(SessionProxyHandler.java:113)
                at com.sun.proxy.$Proxy12.save(Unknown Source)
                at biz.netcentric.cq.tools.actool.installationhistory.impl.AcHistoryServiceImpl.persistHistory(AcHistoryServiceImpl.java:91)
                at biz.netcentric.cq.tools.actool.aceservice.impl.AceServiceImpl.installConfigurationFiles(AceServiceImpl.java:209)
                at biz.netcentric.cq.tools.actool.aceservice.impl.AceServiceImpl.execute(AceServiceImpl.java:134)
                at biz.netcentric.cq.tools.actool.aceservice.impl.AceServiceImpl.execute(AceServiceImpl.java:118)
                at com.foobar.startup.StartupCheckServlet.activate(StartupCheckServlet.java:48)
                at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
…
Caused by: org.apache.jackrabbit.oak.api.CommitFailedException: OakState0001: Unresolved conflicts in /var/statistics/achistory
                at org.apache.jackrabbit.oak.plugins.commit.ConflictValidator.failOnMergeConflict(ConflictValidator.java:115)
                at org.apache.jackrabbit.oak.plugins.commit.ConflictValidator.propertyAdded(ConflictValidator.java:84)
                at org.apache.jackrabbit.oak.spi.commit.CompositeEditor.propertyAdded(CompositeEditor.java:83)
                at org.apache.jackrabbit.oak.spi.commit.EditorDiff.propertyAdded(EditorDiff.java:82)
                at org.apache.jackrabbit.oak.plugins.memory.ModifiedNodeState.compareAgainstBaseState(ModifiedNodeState.java:378)
                at org.apache.jackrabbit.oak.spi.commit.EditorDiff.childNodeChanged(EditorDiff.java:148)
                at org.apache.jackrabbit.oak.plugins.memory.ModifiedNodeState.compareAgainstBaseState(ModifiedNodeState.java:399)
                at org.apache.jackrabbit.oak.spi.commit.EditorDiff.childNodeChanged(EditorDiff.java:148)
                at org.apache.jackrabbit.oak.plugins.memory.ModifiedNodeState.compareAgainstBaseState(ModifiedNodeState.java:399)
                at org.apache.jackrabbit.oak.spi.commit.EditorDiff.childNodeChanged(EditorDiff.java:148)
                at org.apache.jackrabbit.oak.plugins.memory.ModifiedNodeState.compareAgainstBaseState(ModifiedNodeState.java:399)
                at org.apache.jackrabbit.oak.spi.commit.EditorDiff.process(EditorDiff.java:52)
                at org.apache.jackrabbit.oak.spi.commit.EditorHook.processCommit(EditorHook.java:54)
                at org.apache.jackrabbit.oak.spi.commit.CompositeHook.processCommit(CompositeHook.java:61)
                at org.apache.jackrabbit.oak.spi.commit.CompositeHook.processCommit(CompositeHook.java:61)
                at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreBranch$InMemory.merge(DocumentNodeStoreBranch.java:543)
                at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreBranch.merge0(DocumentNodeStoreBranch.java:223)
                at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreBranch.merge(DocumentNodeStoreBranch.java:171)
                at org.apache.jackrabbit.oak.plugins.document.DocumentRootBuilder.merge(DocumentRootBuilder.java:158)
                at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStore.merge(DocumentNodeStore.java:1497)
                at org.apache.jackrabbit.oak.core.MutableRoot.commit(MutableRoot.java:247)
                at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.commit(SessionDelegate.java:339)
                at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.save(SessionDelegate.java:485)
                ... 92 common frames omitted
 ```

--- a/accesscontroltool-bundle/pom.xml
+++ b/accesscontroltool-bundle/pom.xml
@@ -32,6 +32,14 @@
             <artifactId>org.osgi.core</artifactId>
         </dependency>
         <dependency>
+            <groupId>org.apache.sling</groupId>
+            <artifactId>org.apache.sling.event.dea</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.sling</groupId>
+            <artifactId>org.apache.sling.commons.scheduler</artifactId>
+        </dependency>
+        <dependency>
             <groupId>org.apache.felix</groupId>
             <artifactId>org.apache.felix.scr.annotations</artifactId>
         </dependency>

--- a/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configuploadlistener/impl/UploadListenerServiceImpl.java
+++ b/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configuploadlistener/impl/UploadListenerServiceImpl.java
@@ -8,14 +8,10 @@
  */
 package biz.netcentric.cq.tools.actool.configuploadlistener.impl;
 
+import java.util.Date;
+import java.util.Dictionary;
+import java.util.Hashtable;
 import java.util.Map;
-
-import javax.jcr.Node;
-import javax.jcr.RepositoryException;
-import javax.jcr.Session;
-import javax.jcr.observation.Event;
-import javax.jcr.observation.EventIterator;
-import javax.jcr.observation.EventListener;
 
 import org.apache.commons.lang.StringUtils;
 import org.apache.felix.scr.annotations.Activate;
@@ -26,14 +22,20 @@ import org.apache.felix.scr.annotations.Property;
 import org.apache.felix.scr.annotations.PropertyOption;
 import org.apache.felix.scr.annotations.Reference;
 import org.apache.felix.scr.annotations.Service;
+import org.apache.sling.api.SlingConstants;
 import org.apache.sling.commons.osgi.PropertiesUtil;
-import org.apache.sling.jcr.api.SlingRepository;
+import org.apache.sling.commons.scheduler.ScheduleOptions;
+import org.apache.sling.commons.scheduler.Scheduler;
+import org.apache.sling.event.dea.DEAConstants;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.event.EventConstants;
+import org.osgi.service.event.EventHandler;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import biz.netcentric.cq.tools.actool.api.AcInstallationService;
 import biz.netcentric.cq.tools.actool.configuploadlistener.UploadListenerService;
-import biz.netcentric.cq.tools.actool.helper.Constants;
 import biz.netcentric.cq.tools.actool.impl.AcInstallationServiceImpl;
 
 @Component(metatype = true, label = "AC Configuration Upload Listener Service", immediate = true, description = "Listens for ACL configuration uploads and triggers ACL Service.")
@@ -42,63 +44,71 @@ import biz.netcentric.cq.tools.actool.impl.AcInstallationServiceImpl;
 @Property(label = "Service status", name = UploadListenerServiceImpl.ACE_UPLOAD_LISTENER_SET_STATUS_SERVICE, options = {
         @PropertyOption(name = "disabled", value = "disabled"),
         @PropertyOption(name = "enabled", value = "enabled") }) })
+@Property(label = "Execution delay", name = UploadListenerServiceImpl.ACE_UPLOAD_LISTENER_TRIGGER_DELAY_IN_MS, 
+description = "Defer execution of the ACTool after detecting the first changed YAML file by this amount of milliseconds", 
+intValue=UploadListenerServiceImpl.ACE_UPLOAD_LISTENER_TRIGGER_DELAY_IN_MS_DEFAULT_VALUE)
 @Service(value = UploadListenerService.class)
-public class UploadListenerServiceImpl implements UploadListenerService, EventListener {
+public class UploadListenerServiceImpl implements UploadListenerService, EventHandler {
     private static final Logger LOG = LoggerFactory.getLogger(UploadListenerServiceImpl.class);
 
     static final String ACE_UPLOAD_LISTENER_SET_STATUS_SERVICE = "AceUploadListener.setStatusService";
+    static final String ACE_UPLOAD_LISTENER_TRIGGER_DELAY_IN_MS = "AceUploadListener.triggerDelayMs";
 
+    static final int ACE_UPLOAD_LISTENER_TRIGGER_DELAY_IN_MS_DEFAULT_VALUE = 5000;
+    
     private String configurationPath;
     private boolean enabled;
-
-    private Session adminSession;
-
-    @Reference
-    SlingRepository repository;
+    private int executionDelayInMs;
 
     @Reference
     AcInstallationService acInstallationService;
+    
+    @Reference
+    Scheduler scheduler;
+
+    private ServiceRegistration eventHandlerServiceRegistration;
 
     @Override
-    public void onEvent(EventIterator events) {
-
-        if (this.enabled) {
-            int changes = 0; // Number of new or changed files.
-            try {
-                while (events.hasNext()) {
-                    Event event = events.nextEvent();
-                    Node node = null;
-                    switch (event.getType()) {
-                    case Event.NODE_ADDED:
-                        node = adminSession.getNode(event.getPath());
-                        break;
-                    case Event.PROPERTY_CHANGED:
-                        if (event.getPath().endsWith("jcr:content/jcr:data")) {
-                            node = adminSession.getNode(event.getPath().replace("/jcr:content/jcr:data", ""));
-                        }
-                        break;
-                    default:
-                        LOG.warn("Unexpected event: {}", event);    
+    public void handleEvent(org.osgi.service.event.Event event) {
+        LOG.debug("UploadListener Event Handler triggered for event '{}'", event);
+        // check if triggered from local instance (in case of clustered environment)
+        // http://sling.apache.org/documentation/the-sling-engine/resources.html#osgi-event-admin
+        if (event.getProperty(DEAConstants.PROPERTY_APPLICATION) != null) {
+            LOG.debug("Ignore remotely triggered resource event");
+            return;
+        }
+        // extract property path
+        String path = (String)event.getProperty(SlingConstants.PROPERTY_PATH);
+        // if it ends with .yaml and has resource type nt:file
+        if (path != null && path.endsWith(".yaml")) {
+            // check resource type
+            String resourceType = (String)event.getProperty(SlingConstants.PROPERTY_RESOURCE_TYPE);
+            if (resourceType != null && resourceType.equals("nt:file")) 
+            {
+                // most probably an updated yaml file
+                LOG.info("There are new or changed YAML files. Triggering execution of actool with a delay of {} milliseconds.", executionDelayInMs);
+                // we need to trigger the installation in a new thread, otherwise it may be blacklisted (http://felix.apache.org/documentation/subprojects/apache-felix-event-admin.html)
+                // also delay processing until all YAML files have been placed there
+                final ScheduleOptions options = scheduler.AT(new Date(System.currentTimeMillis() + executionDelayInMs));
+                options.name("UploadListenerServiceImpl: trigger installation service");
+                scheduler.schedule(new Runnable() {
+                    @Override
+                    public void run() {
+                        LOG.info("Triggering reload of configuration.");
+                        acInstallationService.apply();
                     }
-                    if (node != null && node.hasProperty("jcr:content/jcr:data")) {
-                        LOG.info("Detected new or changed node at {}", node.getPath());
-                        changes++;
-                    } else {
-                        LOG.debug("Node {} associated with event does not have configuration data.", event.getPath());
-                    }
-                }
-            } catch (RepositoryException e) {
-                LOG.error("Error while handling events.", e);
+                }, options);
+            } else {
+                LOG.debug("Observed resource event is not of resource type nt:file but of resource type '{}'", resourceType);
             }
-            if (changes > 0) {
-                LOG.info("There are {} new or changed files. Triggering reload of configuration.", changes);
-                acInstallationService.apply();
-            }
+        } else {
+            LOG.debug("Observed resource event path is not ending with '.yaml': '{}'", path);
         }
     }
 
+    @SuppressWarnings("unchecked")
     @Activate
-    public void activate(@SuppressWarnings("rawtypes") final Map properties)
+    public void activate(BundleContext bundleContext, @SuppressWarnings("rawtypes") final Map properties)
             throws Exception {
         this.configurationPath = ((AcInstallationServiceImpl) acInstallationService).getConfiguredAcConfigurationRootPath();
         String statusService = PropertiesUtil.toString(properties.get(UploadListenerServiceImpl.ACE_UPLOAD_LISTENER_SET_STATUS_SERVICE),
@@ -108,7 +118,7 @@ public class UploadListenerServiceImpl implements UploadListenerService, EventLi
         } else {
             this.enabled = false;
         }
-
+        this.executionDelayInMs = PropertiesUtil.toInteger(properties.get(ACE_UPLOAD_LISTENER_TRIGGER_DELAY_IN_MS), ACE_UPLOAD_LISTENER_TRIGGER_DELAY_IN_MS_DEFAULT_VALUE);
         if (!this.enabled) {
             LOG.debug("UploadListenerServiceImpl is not active, not registering listener");
             return;
@@ -117,48 +127,28 @@ public class UploadListenerServiceImpl implements UploadListenerService, EventLi
                     + "biz.netcentric.cq.tools.actool.impl.AcInstallationServiceImpl/'AceService.configurationPath' to be configured");
             return;
         } else {
-            setupEventListener();
-        }
-    }
-
-    private void setupEventListener() throws Exception {
-        try {
-            adminSession = repository.loginService(Constants.USER_AC_SERVICE, null);
-
-            adminSession
-                    .getWorkspace()
-                    .getObservationManager()
-                    .addEventListener(
-
-                            this, // handler
-
-                            // Event.PROPERTY_ADDED|Event.NODE_ADDED,
-                            // //binary combination of event types
-                            Event.NODE_ADDED | Event.PROPERTY_CHANGED,
-                            this.configurationPath, // path
-
-                            true, // is Deep?
-
-                            null, // uuids filter
-
-                            null, // nodetypes filter
-                            false);
-            LOG.info("Registered event listener for AC configuration root path: {}", this.configurationPath);
-        } catch (RepositoryException e) {
-            LOG.error("Exception while registering listener in UploadListenerService: " + e, e);
+            // subscribe to resource events via OSGi Event Admin
+            // http://sling.apache.org/documentation/the-sling-engine/resources.html#osgi-event-admin
+            @SuppressWarnings("rawtypes")
+            Dictionary eventHandlerProperties = new Hashtable();
+            String[] topics = new String[] {SlingConstants.TOPIC_RESOURCE_ADDED, SlingConstants.TOPIC_RESOURCE_CHANGED};
+            eventHandlerProperties.put(EventConstants.EVENT_TOPIC, topics);
+            // make sure we only listen for a specific root path
+            eventHandlerProperties.put(EventConstants.EVENT_FILTER, String.format("(%s=%s/*)", SlingConstants.PROPERTY_PATH, configurationPath));
+            eventHandlerServiceRegistration = bundleContext.registerService(EventHandler.class.getName(), this, eventHandlerProperties);
+            LOG.info("Registered event handler for AC configuration root path: {}", this.configurationPath);
         }
     }
 
     @Deactivate
     public void deactivate() {
-        if (adminSession != null) {
+        if (eventHandlerServiceRegistration != null) {
             try {
-                adminSession.getWorkspace().getObservationManager().removeEventListener(this);
-                LOG.info("Unregistered event listener for AC configuration root path: {}", this.configurationPath);
+                eventHandlerServiceRegistration.unregister();
+                LOG.info("Unregistered event handler for AC configuration root path: {}", this.configurationPath);
             } catch (Exception e) {
-                LOG.error("Exception while unregistering listener in UploadListenerService: " + e, e);
+                LOG.error("Exception while unregistering event handler in UploadListenerService: " + e, e);
             }
-            adminSession.logout();
         }
     }
 

--- a/pom.xml
+++ b/pom.xml
@@ -82,7 +82,7 @@
 
     <dependencyManagement>
         <dependencies>
-            <!-- Runtime dependencies should be CQ 5.6.1 (SP2) compatible -->
+            <!-- Runtime dependencies should be AEM 6.1 compatible -->
             <dependency>
                 <groupId>org.apache.jackrabbit</groupId>
                 <artifactId>jackrabbit-jcr-commons</artifactId>
@@ -117,6 +117,21 @@
                 <groupId>org.osgi</groupId>
                 <artifactId>org.osgi.compendium</artifactId>
                 <version>4.2.0</version>
+                <scope>provided</scope>
+            </dependency>
+            <!-- only constants are referenced from there which are inlined during compilation anyways, therefore no runtime dependency -->
+            <dependency>
+                <groupId>org.apache.sling</groupId>
+                <artifactId>org.apache.sling.event.dea</artifactId>
+                <!-- shipped with AEM 6.1 -->
+                <version>1.0.0</version>
+                <scope>provided</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.apache.sling</groupId>
+                <artifactId>org.apache.sling.commons.scheduler</artifactId>
+                <!-- shipped with AEM 6.1 -->
+                <version>2.4.6</version>
                 <scope>provided</scope>
             </dependency>
             <dependency>

